core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -859,16 +871,19 @@ class KafkaApis(val requestChannel: RequestChannel,
                     _ : UnsupportedForMessageFormatException) =>
             debug(s"Offset request with correlation id $correlationId from client $clientId on " +
                 s"partition $topicPartition failed due to ${e.getMessage}")
-            (topicPartition, new ListOffsetResponse.PartitionData(Errors.forException(e),
-              ListOffsetResponse.UNKNOWN_TIMESTAMP,
-              ListOffsetResponse.UNKNOWN_OFFSET,
-              Optional.empty()))
+            buildErrorResponse(Errors.forException(e))
+
+          // Only V5 and newer ListOffset calls should get OFFSET_NOT_AVAILABLE
+          case e: OffsetNotAvailableException =>
+            if(isV5Schema) {
+              buildErrorResponse(Errors.forException(e))
```

Comment:
Just thinking about this a little, do you think there's much advantage to having a separate error code? Whether it is `OFFSET_NOT_AVAILABLE` or `LEADER_NOT_AVAILABLE`, the client will just retry.

Also, it might be a good idea to update `Fetcher.handleListOffsetResponse` to explicitly check the expected error codes. We can probably skip the log warning in these cases and print a debug message since this is an expected scenario.

Reply:
The motivation of a new exception was to give a more precise reason for not returning the offsets. Sticking with `LEADER_NOT_AVAILABLE` would eliminate any change required to the protocol, which is nice, but maybe runs the risk of overloading the error code. Looking at usages of `LEADER_NOT_AVAILABLE` in the code, it doesn't actually look all that widely used, so maybe adding another usage of it for this case isn't so bad. 

@hachikuji @cmccabe WDYT?

Reply:
> Fetcher.handleListOffsetResponse to explicitly check the expected error codes

Sounds good. Would handling `LEADER_NOT_AVAILABLE` and retrying have any unintended consequences with other uses of this error code?

Reply:
Yeah, I'm not sure. We have tended to regret when we didn't provide more explicit error codes, so perhaps we should just stick with what's in the KIP. 

> Sounds good. Would handling LEADER_NOT_AVAILABLE and retrying have any unintended consequences with other uses of this error code?

I didn't see any explicit logic to handle it in `Fetcher`, so I think we'd be hitting this case:
```java
log.warn("Attempt to fetch offsets for partition {} failed due to: {}", topicPartition, error.message());
partitionsToRetry.add(topicPartition);
```
Do we have any current scenarios in the ListOffset handler where we could raise `LEADER_NOT_AVAILABLE`?

Reply:
As far as I can tell, `LEADER_NOT_AVAILABLE` is only returned when creating a topic, fetching metadata, and deleting records. I'll continue with the new error as indicated in the KIP and rev the IBP

Reply:
Yes, we discussed this in the KIP.  It is good to be specific with error codes.  +1.